esphome/components/fastled/light/__init__.py

```diff
@@ -0,0 +1,163 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome import pins
+from esphome.components import light
+from esphome.const import (
+    CONF_CHIPSET,
+    CONF_CLOCK_PIN,
+    CONF_DATA_PIN,
+    CONF_DATA_RATE,
+    CONF_PIN,
+    CONF_RGB_ORDER,
+    CONF_OUTPUT_ID,
+    CONF_NUM_LEDS,
+    CONF_MAX_REFRESH_RATE,
+)
+
+CODEOWNERS = ["@OttoWinter", "@NielsNL68"]
+
+fastled_base_ns = cg.esphome_ns.namespace("fastled")
+FastLEDLightOutput = fastled_base_ns.class_(
+    "FastLEDLightOutput", light.AddressableLight
+)
+
+RGB_ORDERS = [
+    "RGB",
+    "RBG",
+    "GRB",
+    "GBR",
+    "BRG",
+    "BGR",
+]
+
+SPI_CHIPSETS = [
+    "LPD6803",
+    "LPD8806",
+    "WS2801",
+    "WS2803",
+    "SM16716",
+    "P9813",
+    "APA102",
+    "SK9822",
+    "DOTSTAR",
+]
+
+CLOCKLESS_CHIPSETS = [
+    "NEOPIXEL",
+    "TM1829",
+    "TM1809",
+    "TM1804",
+    "TM1803",
+    "UCS1903",
+    "UCS1903B",
+    "UCS1904",
+    "UCS2903",
+    "WS2812",
+    "WS2852",
+    "WS2812B",
+    "SK6812",
+    "SK6822",
+    "APA106",
+    "PL9823",
+    "WS2811",
+    "WS2813",
+    "APA104",
+    "WS2811_400",
+    "GW6205",
+    "GW6205_400",
+    "LPD1886",
+    "LPD1886_8BIT",
+    "SM16703",
+]
+
+CHIPSETS = CLOCKLESS_CHIPSETS + SPI_CHIPSETS
+
+
+def _validate(config):
+    if config[CONF_CHIPSET] == "NEOPIXEL" and CONF_RGB_ORDER in config:
+        raise cv.Invalid("NEOPIXEL doesn't support RGB order")
+    if config[CONF_CHIPSET] in SPI_CHIPSETS and config[CONF_CLOCK_PIN] == -1:
+        raise cv.Invalid("The clock_pin is required for SPI devices.")
+
+    return config
+
+
+def validate_gpio_output_pin_number(value):
+    if value == -1:
+        return value
+    return pins.internal_gpio_output_pin_number(value)
+
+
+BASE_SCHEMA = light.ADDRESSABLE_LIGHT_SCHEMA.extend(
+    {
+        cv.GenerateID(CONF_OUTPUT_ID): cv.declare_id(FastLEDLightOutput),
+        cv.Required(CONF_NUM_LEDS): cv.positive_not_null_int,
+        cv.Optional(CONF_RGB_ORDER): cv.one_of(*RGB_ORDERS, upper=True),
+        cv.Optional(CONF_MAX_REFRESH_RATE): cv.positive_time_period_microseconds,
+    }
+).extend(cv.COMPONENT_SCHEMA)
+
+
+CONFIG_SCHEMA = cv.All(
+    BASE_SCHEMA.extend(
+        {
+            cv.GenerateID(CONF_OUTPUT_ID): cv.declare_id(FastLEDLightOutput),
+            cv.Required(CONF_CHIPSET): cv.one_of(*CHIPSETS, upper=True),
+            cv.Required(CONF_NUM_LEDS): cv.positive_not_null_int,
+            cv.Optional(CONF_RGB_ORDER): cv.one_of(*RGB_ORDERS, upper=True),
+            cv.Optional(CONF_MAX_REFRESH_RATE): cv.positive_time_period_microseconds,
+            cv.Optional(CONF_DATA_PIN): pins.internal_gpio_output_pin_number,
+            cv.Optional(CONF_PIN): cv.Invalid("This pin is renamed to 'data_pin'"),
```

Comment:
When you raise an error, it's a capitalised `Invalid`, but in the schema, it's lower case `invalid`
```suggestion
            cv.Optional(CONF_PIN): cv.invalid("This pin is renamed to 'data_pin'"),
```

Reply:
then i copied the line from the wrong place. thanks for you note.